File: clients/js/src/features/wrap_feature_env_vars.ts
import { Input } from "../input";

import { PolywrapClientConfigBuilder, PolywrapClient, Uri } from "@polywrap/client-js";
import path from "path";

export async function runTestCase(input: unknown): Promise<void> {
  const inputObj = Input.expectObject<{
    mainEnv: unknown;
    subinvokerEnv: unknown;
  }>(input);

  const mainEnv = Input.expectObject<Record<string, unknown>>(
    inputObj.mainEnv
  );
  const subinvokerEnv = Input.expectObject<Record<string, unknown>>(
    inputObj.subinvokerEnv
  );

  const wrapsDir = path.join(__dirname, "../../../../wraps");
  const mainPath = path.join(wrapsDir, "/env-type/00-main/implementations/as");
  const mainUri = Uri.from("file/" + mainPath);
  const subinvokerPath = path.join(wrapsDir, "/env-type/02-subinvoker-with-env/implementations/as");
  const subinvokerUri = Uri.from("file/" + subinvokerPath);

  const envs = {
    [mainUri.uri]: mainEnv,
    [subinvokerUri.uri]: subinvokerEnv,
  };

  const config = new PolywrapClientConfigBuilder()
    .addDefaults()
    .addEnvs(envs)
    .addRedirect("mock/main", mainUri.uri)
    .build();

  const client = new PolywrapClient(config);

  console.log("Invoking methodRequireEnv");

  const methodRequireEnvResult = await client.invoke<{
    str: string
  }>({
    uri: mainUri.uri,
    method: "methodRequireEnv",
    args: {
      arg: "string",
    },
  });

  if (!methodRequireEnvResult.ok) throw methodRequireEnvResult.error;
  console.log(
    "response.str:", methodRequireEnvResult.value.str
  );

  console.log("Success!");

  console.log("Invoking subinvokeMethodRequireEnv");

  const subinvokeEnvMethodResult = await client.invoke<{
    str: string
  }>({
    uri: subinvokerUri.uri,
    method: "subinvokeMethodRequireEnv",
    args: {
      arg: "string",
    },
  });

  if (!subinvokeEnvMethodResult.ok) throw subinvokeEnvMethodResult.error;

  console.log(
    "response.str:", subinvokeEnvMethodResult.value.str
  );
  console.log("Success!")
}

File: clients/py/src/features/wrap_feature_env_vars.py
from typing import Any, Dict, Optional, TypedDict, List
from pydantic import BaseModel, Field
from polywrap_client import PolywrapClient
from polywrap_client_config_builder import PolywrapClientConfigBuilder
from polywrap_uri_resolvers import SimpleFileReader, FsUriResolver
from polywrap_core import Uri
from pathlib import Path

Env = TypedDict(
    "Env",
    {
        "str": str,
        "object": Any,
        "array": List[int],
        "number": int,
        "bool": bool,
        "en": str,
        "optFilledStr": Optional[str],
    },
    total=False,
)

ExtEnv = TypedDict(
    "ExtEnv",
    {
        "externalArray": List[int],
        "externalString": str,
    },
    total=False,
)


class TestCaseInput(BaseModel):
    main_env: Env = Field(..., alias="mainEnv")
    subinvoker_env: Env = Field(..., alias="subinvokerEnv")


def run_test_case(input: Any) -> None:
    input_obj = TestCaseInput.parse_obj(input)

    root = Path(__file__).parent.parent.parent.parent.parent / "wraps"
    external_wrapper_path = root / "env-type/00-external/implementations/as"
    external_wrapper_uri = Uri.from_str(f"fs/{external_wrapper_path}")

    wrapper_path = root / "env-type/01-main/implementations/as"
    wrapper_uri = Uri.from_str(f"fs/{wrapper_path}")

    subinvoker_path = root / "env-type/02-subinvoker-with-env/implementations/as"
    subinvoker_uri = Uri.from_str(f"fs/{subinvoker_path}")

    envs = {
        wrapper_uri: input_obj.main_env,
        external_wrapper_uri: input_obj.subinvoker_env,
    }

    config = (
        PolywrapClientConfigBuilder()
        .set_envs(envs)
        .set_redirect(
            Uri.from_str("ens/external-env.polywrap.eth"), external_wrapper_uri
        )
        .add_resolver(FsUriResolver(SimpleFileReader()))
        .build()
    )

    client = PolywrapClient(config)

    print("Invoking methodRequireEnv")

    method_require_env_result = client.invoke(
        uri=wrapper_uri,
        method="methodRequireEnv",
        args={
            "arg": "string",
        },
    )

    if not method_require_env_result:
        raise Exception(f"Error: {method_require_env_result}")

    print("response.str:", method_require_env_result.get("str"))
    print("Success!")

    print("Invoking subinvokeMethodRequireEnv")

    print(subinvoker_uri)

    subinvoke_env_method_result = client.invoke(
        uri=subinvoker_uri,
        method="subinvokeMethodRequireEnv",
        args={
            "arg": "string",
        },
    )

    print(subinvoke_env_method_result)

    if not subinvoke_env_method_result:
        raise Exception(f"Error: {subinvoke_env_method_result}")

    print("response.str:", subinvoke_env_method_result.get("str"))
    print("Success!")
